packages/@aws-cdk/custom-resources/lib/provider-framework/runtime/framework.ts

```diff
@@ -0,0 +1,177 @@
+  });
+}
+
+// invoked a few times until `complete` is true or until it times out.
+async function isComplete(event: AWSCDKAsyncCustomResource.IsCompleteRequest) {
+  return await failOnError(event, async () => {
+    log('isComplete', event);
+
+    const isCompleteResult = await invokeUserFunction(consts.USER_IS_COMPLETE_FUNCTION_ARN_ENV, event) as IsCompleteResponse;
+    log('user isComplete returned:', isCompleteResult);
+
+    // if we are not complete, reeturn false, and don't send a response back.
+    if (!isCompleteResult.IsComplete) {
+      if (isCompleteResult.Data && Object.keys(isCompleteResult.Data).length > 0) {
+        throw new Error(`"Data" is not allowed if "IsComplete" is "False"`);
+      }
+
+      throw new Retry(JSON.stringify(event));
+    }
+
+    const response = {
+      ...event,
+      Data: {
+        ...event.Data,
+        ...isCompleteResult.Data
+      }
+    };
+
+    await submitCloudFormationResponse('SUCCESS', response);
+  });
+}
+
+// invoked when completion retries are exhaused.
+async function onTimeout(timeoutEvent: any) {
+  log('timeoutHandler', timeoutEvent);
+
+  const isCompleteRequest = JSON.parse(JSON.parse(timeoutEvent.Cause).errorMessage) as AWSCDKAsyncCustomResource.IsCompleteRequest;
+  await submitCloudFormationResponse('FAILED', isCompleteRequest, {
+    reason: 'Operation timed out'
+  });
+}
+
+async function invokeUserFunction(arnEnvironment: string, payload: any) {
+  const arn = getEnv(arnEnvironment);
+  log(`executing user function ${arnEnvironment} with payload`, payload);
+  const resp = await invokeFunction({ FunctionName: arn, Payload: JSON.stringify(payload) });
+
+  log('user function response:', resp, typeof(resp));
+
+  const jsonPayload = parseJsonPayload(resp.Payload);
+  if (resp.FunctionError) {
+    log('user function threw an error:', resp.FunctionError);
+    const errorMessage = jsonPayload.errorMessage || 'error';
+    const trace = jsonPayload.trace ? `\nRemote function error: ` + jsonPayload.trace.join('\n') : '';
+
+    const e = new Error(errorMessage);
+    e.stack += trace;
+    throw e;
+  }
+
+  return jsonPayload;
+}
+
+function parseJsonPayload(payload: any): any {
+  if (!payload) { return { }; }
+  const text = payload.toString();
+  try {
+    return JSON.parse(text);
+  } catch (e) {
+    log('parse error:', e);
+    return { text };
+  }
+}
+
+function validateCfnRequest(cfnRequest: AWSLambda.CloudFormationCustomResourceEvent) {
+  const errorPrefix = `Invalid CloudFormation custom resource event (${cfnRequest.RequestType || ''}):`;
+
+  [ 'LogicalResourceId', 'RequestId', 'RequestType', 'ResponseURL', 'StackId', 'ServiceToken' ].forEach(field => {
+    if (!(cfnRequest as any)[field]) {
+      throw new Error(`${errorPrefix} ${field} is required`);
+    }
+  });
+
+  if (cfnRequest.RequestType === 'Create' && (cfnRequest as any).PhysicalResourceId) {
+    throw new Error(`${errorPrefix} PhysicalResourceId is not allowed for "Create" events`);
```

Comment:
Why are you testing CFNs calls to you here?

Reply:
Defensiveness I guess, but probably redudant.

Reply:
removed.